File: server/routes/teams.js
const express = require('express');
const router = express.Router();
const { query, cache } = require('../database/connection');

// GET /api/v1/teams - 모든 팀 목록
router.get('/', async (req, res) => {
    try {
        const cacheKey = 'teams:all';
        
        // Check cache first
        let cachedData = await cache.get(cacheKey);
        if (cachedData) {
            return res.json({
                success: true,
                data: cachedData,
                source: 'cache',
                timestamp: new Date().toISOString()
            });
        }

        const teamsQuery = `
            SELECT 
                team_id,
                team_name,
                team_name_en,
                team_name_jp,
                team_abbreviation,
                league,
                city,
                stadium,
                established_year,
                team_color,
                logo_url,
                website_url,
                is_active
            FROM teams 
            WHERE is_active = true
            ORDER BY league, team_name
        `;

        const result = await query(teamsQuery);
        
        // Group by league
        const teams = {
            central: result.rows.filter(team => team.league === 'central'),
            pacific: result.rows.filter(team => team.league === 'pacific'),
            total: result.rows.length
        };

        // Cache for 1 hour (teams data doesn't change often)
        await cache.set(cacheKey, teams, 3600);

        res.json({
            success: true,
            data: teams,
            source: 'database',
            timestamp: new Date().toISOString()
        });

    } catch (error) {
        console.error('Teams API error:', error);
        res.status(500).json({
            success: false,
            error: 'Failed to fetch teams data',
            code: 'TEAMS_FETCH_ERROR'
        });
    }
});

// GET /api/v1/teams/:league - 특정 리그 팀 목록
router.get('/:league', async (req, res) => {
    try {
        const { league } = req.params;
        
        if (!['central', 'pacific'].includes(league)) {
            return res.status(400).json({
                success: false,
                error: 'Invalid league. Must be "central" or "pacific"',
                code: 'INVALID_LEAGUE'
            });
        }

        const cacheKey = `teams:league:${league}`;
        
        // Check cache first
        let cachedData = await cache.get(cacheKey);
        if (cachedData) {
            return res.json({
                success: true,
                data: cachedData,
                source: 'cache',
                timestamp: new Date().toISOString()
            });
        }

        const teamsQuery = `
            SELECT 
                team_id,
                team_name,
                team_name_en,
                team_name_jp,
                team_abbreviation,
                league,
                city,
                stadium,
                established_year,
                team_color,
                logo_url,
                website_url,
                is_active
            FROM teams 
            WHERE league = $1 AND is_active = true
            ORDER BY team_name
        `;

        const result = await query(teamsQuery, [league]);
        
        const teams = {
            league: league,
            teams: result.rows,
            count: result.rows.length
        };

        // Cache for 1 hour
        await cache.set(cacheKey, teams, 3600);

        res.json({
            success: true,
            data: teams,
            source: 'database',
            timestamp: new Date().toISOString()
        });

    } catch (error) {
        console.error('League teams API error:', error);
        res.status(500).json({
            success: false,
            error: 'Failed to fetch league teams data',
            code: 'LEAGUE_TEAMS_FETCH_ERROR'
        });
    }
});

// GET /api/v1/teams/team/:teamId - 특정 팀 상세 정보
router.get('/team/:teamId', async (req, res) => {
    try {
        const { teamId } = req.params;
        
        if (!teamId || isNaN(teamId)) {
            return res.status(400).json({
                success: false,
                error: 'Invalid team ID',
                code: 'INVALID_TEAM_ID'
            });
        }

        const cacheKey = `teams:detail:${teamId}`;
        
        // Check cache first
        let cachedData = await cache.get(cacheKey);
        if (cachedData) {
            return res.json({
                success: true,
                data: cachedData,
                source: 'cache',
                timestamp: new Date().toISOString()
            });
        }

        // Get team details with current season stats
        const currentYear = new Date().getFullYear();
        const teamDetailQuery = `
            SELECT 
                t.*,
                s.rank,
                s.wins,
                s.losses,
                s.draws,
                s.games_played,
                s.win_percentage,
                s.games_behind,
                s.runs_scored,
                s.runs_allowed,
                s.run_differential,
                s.streak_type,
                s.streak_count,
                s.home_wins,
                s.home_losses,
                s.home_draws,
                s.away_wins,
                s.away_losses,
                s.away_draws,
                s.last10_wins,
                s.last10_losses,
                s.last10_draws,
                s.magic_number,
                s.elimination_number,
                s.last_updated as standings_updated
            FROM teams t
            LEFT JOIN standings s ON t.team_id = s.team_id AND s.season = $2
            WHERE t.team_id = $1 AND t.is_active = true
        `;

        const result = await query(teamDetailQuery, [teamId, currentYear]);
        
        if (result.rows.length === 0) {
            return res.status(404).json({
                success: false,
                error: 'Team not found',
                code: 'TEAM_NOT_FOUND'
            });
        }

        const teamData = result.rows[0];

        // Get recent games (last 5)
        const recentGamesQuery = `
            SELECT 
                g.game_id,
                g.game_date,
                g.game_status,
                g.home_score,
                g.away_score,
                g.innings,
                g.is_extra_innings,
                CASE 
                    WHEN g.home_team_id = $1 THEN 'home'
                    ELSE 'away'
                END as venue,
                CASE 
                    WHEN g.home_team_id = $1 THEN away_team.team_name
                    ELSE home_team.team_name
                END as opponent_name,
                CASE 
                    WHEN g.home_team_id = $1 THEN away_team.team_abbreviation
                    ELSE home_team.team_abbreviation
                END as opponent_abbr,
                CASE 
                    WHEN g.home_team_id = $1 THEN away_team.team_color
                    ELSE home_team.team_color
                END as opponent_color,
                CASE 
                    WHEN g.game_status = 'completed' THEN
                        CASE 
                            WHEN (g.home_team_id = $1 AND g.home_score > g.away_score) OR
                                 (g.away_team_id = $1 AND g.away_score > g.home_score) THEN 'W'
                            WHEN g.home_score = g.away_score THEN 'T'
                            ELSE 'L'
                        END
                    ELSE null
                END as result
            FROM games g
            JOIN teams home_team ON g.home_team_id = home_team.team_id
            JOIN teams away_team ON g.away_team_id = away_team.team_id
            WHERE (g.home_team_id = $1 OR g.away_team_id = $1)
                AND g.game_date <= CURRENT_DATE
            ORDER BY g.game_date DESC, g.game_id DESC
            LIMIT 5
        `;

        const recentGamesResult = await query(recentGamesQuery, [teamId]);

        // Calculate additional metrics
        const teamDetail = {
            ...teamData,
            season: currentYear,
            recent_games: recentGamesResult.rows,
            calculated_metrics: {
                home_win_percentage: teamData.home_wins + teamData.home_losses > 0 ? 
                    Math.round((teamData.home_wins / (teamData.home_wins + teamData.home_losses)) * 1000) / 1000 : 0,
                away_win_percentage: teamData.away_wins + teamData.away_losses > 0 ? 
                    Math.round((teamData.away_wins / (teamData.away_wins + teamData.away_losses)) * 1000) / 1000 : 0,
                last10_win_percentage: teamData.last10_wins + teamData.last10_losses > 0 ? 
                    Math.round((teamData.last10_wins / (teamData.last10_wins + teamData.last10_losses)) * 1000) / 1000 : 0,
                runs_per_game: teamData.games_played > 0 ? 
                    Math.round((teamData.runs_scored / teamData.games_played) * 100) / 100 : 0,
                runs_allowed_per_game: teamData.games_played > 0 ? 
                    Math.round((teamData.runs_allowed / teamData.games_played) * 100) / 100 : 0,
                pythagorean_expectation: teamData.runs_scored > 0 && teamData.runs_allowed > 0 ? 
                    Math.round((Math.pow(teamData.runs_scored, 2) / (Math.pow(teamData.runs_scored, 2) + Math.pow(teamData.runs_allowed, 2))) * 1000) / 1000 : 0
            },
            current_streak: teamData.streak_type && teamData.streak_count > 0 ? 
                `${teamData.streak_type}${teamData.streak_count}` : null
        };

        // Cache for 10 minutes
        await cache.set(cacheKey, teamDetail, 600);

        res.json({
            success: true,
            data: teamDetail,
            source: 'database',
            timestamp: new Date().toISOString()
        });

    } catch (error) {
        console.error('Team detail API error:', error);
        res.status(500).json({
            success: false,
            error: 'Failed to fetch team detail data',
            code: 'TEAM_DETAIL_FETCH_ERROR'
        });
    }
});

module.exports = router;

File: api/routes/teams.py
"""
Teams API Routes
팀 관련 API 엔드포인트
"""

from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional
from ..models.database import db

router = APIRouter(prefix="/teams", tags=["teams"])

@router.get("/")
async def get_all_teams():
    """전체 팀 목록"""
    try:
        teams = db.execute_query("""
            SELECT 
                team_id,
                team_name,
                team_abbreviation,
                league,
                team_color,
                stadium_name,
                founded_year,
                created_at
            FROM teams
            ORDER BY league, team_name
        """)
        
        return {
            "teams": teams,
            "central_league": [t for t in teams if t['league'] == 'Central'],
            "pacific_league": [t for t in teams if t['league'] == 'Pacific'],
            "total_teams": len(teams)
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/{team_id}")
async def get_team_detail(team_id: int):
    """팀 상세 정보"""
    try:
        team = db.execute_single("""
            SELECT 
                t.team_id,
                t.team_name,
                t.team_abbreviation,
                t.league,
                t.team_color,
                t.stadium_name,
                t.founded_year,
                s.position_rank,
                s.games_played,
                s.wins,
                s.losses,
                s.draws,
                s.win_percentage,
                s.games_behind,
                s.runs_scored,
                s.runs_allowed,
                s.run_differential,
                s.home_wins,
                s.home_losses,
                s.away_wins,
                s.away_losses
            FROM teams t
            LEFT JOIN standings s ON t.team_id = s.team_id AND s.season_year = 2025
            WHERE t.team_id = %s
        """, (team_id,))
        
        if not team:
            raise HTTPException(status_code=404, detail="Team not found")
        
        return team
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/{team_id}/games")
async def get_team_games(
    team_id: int,
    limit: int = Query(20, description="조회할 경기 수"),
    home_only: bool = Query(False, description="홈 경기만"),
    away_only: bool = Query(False, description="원정 경기만")
):
    """팀 경기 일정/결과"""
    try:
        query = """
            SELECT 
                g.game_id,
                g.game_date,
                ht.team_abbreviation as home_team,
                at.team_abbreviation as away_team,
                g.home_score,
                g.away_score,
                g.game_status,
                g.is_extra_innings,
                g.is_draw,
                g.stadium,
                CASE 
                    WHEN g.home_team_id = %s THEN 'home'
                    ELSE 'away'
                END as venue,
                CASE 
                    WHEN g.game_status = 'completed' THEN
                        CASE 
                            WHEN (g.home_team_id = %s AND g.home_score > g.away_score) OR
                                 (g.away_team_id = %s AND g.away_score > g.home_score) THEN 'W'
                            WHEN g.is_draw THEN 'D'
                            ELSE 'L'
                        END
                    ELSE null
                END as result
            FROM games g
            JOIN teams ht ON g.home_team_id = ht.team_id
            JOIN teams at ON g.away_team_id = at.team_id
            WHERE (g.home_team_id = %s OR g.away_team_id = %s)
        """
        params = [team_id, team_id, team_id, team_id, team_id]
        
        if home_only:
            query += " AND g.home_team_id = %s"
            params.append(team_id)
        elif away_only:
            query += " AND g.away_team_id = %s"
            params.append(team_id)
        
        query += " ORDER BY g.game_date DESC LIMIT %s"
        params.append(limit)
        
        games = db.execute_query(query, tuple(params))
        
        return {
            "team_id": team_id,
            "games": games,
            "total_games": len(games),
            "record": {
                "wins": len([g for g in games if g['result'] == 'W']),
                "losses": len([g for g in games if g['result'] == 'L']),
                "draws": len([g for g in games if g['result'] == 'D'])
            }
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/{team_id}/vs/{opponent_id}")
async def get_head_to_head(team_id: int, opponent_id: int):
    """팀간 상대전적"""
    try:
        # 상대전적
        h2h_games = db.execute_query("""
            SELECT 
                g.game_id,
                g.game_date,
                ht.team_abbreviation as home_team,
                at.team_abbreviation as away_team,
                g.home_score,
                g.away_score,
                g.game_status,
                g.is_extra_innings,
                g.is_draw,
                CASE 
                    WHEN g.game_status = 'completed' THEN
                        CASE 
                            WHEN (g.home_team_id = %s AND g.home_score > g.away_score) OR
                                 (g.away_team_id = %s AND g.away_score > g.home_score) THEN 'W'
                            WHEN g.is_draw THEN 'D'
                            ELSE 'L'
                        END
                    ELSE null
                END as result_for_team
            FROM games g
            JOIN teams ht ON g.home_team_id = ht.team_id
            JOIN teams at ON g.away_team_id = at.team_id
            WHERE ((g.home_team_id = %s AND g.away_team_id = %s) OR
                   (g.home_team_id = %s AND g.away_team_id = %s))
            ORDER BY g.game_date DESC
        """, (team_id, team_id, team_id, opponent_id, opponent_id, team_id))
        
        # 통계 요약
        wins = len([g for g in h2h_games if g['result_for_team'] == 'W'])
        losses = len([g for g in h2h_games if g['result_for_team'] == 'L'])
        draws = len([g for g in h2h_games if g['result_for_team'] == 'D'])
        
        # 팀 정보
        team_info = db.execute_single("""
            SELECT team_name, team_abbreviation FROM teams WHERE team_id = %s
        """, (team_id,))
        
        opponent_info = db.execute_single("""
            SELECT team_name, team_abbreviation FROM teams WHERE team_id = %s
        """, (opponent_id,))
        
        return {
            "team": team_info,
            "opponent": opponent_info,
            "head_to_head": {
                "wins": wins,
                "losses": losses,
                "draws": draws,
                "total_games": len(h2h_games),
                "win_percentage": wins / len(h2h_games) if h2h_games else 0
            },
            "recent_games": h2h_games[:10]  # 최근 10경기
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/{team_id}/stats")
async def get_team_stats(
    team_id: int,
    season: int = Query(2025, description="시즌 연도")
):
    """팀 상세 통계"""
    try:
        # 기본 통계
        basic_stats = db.execute_single("""
            SELECT 
                s.*,
                t.team_name,
                t.team_abbreviation,
                t.league
            FROM standings s
            JOIN teams t ON s.team_id = t.team_id
            WHERE s.team_id = %s AND s.season_year = %s
        """, (team_id, season))
        
        if not basic_stats:
            raise HTTPException(status_code=404, detail="Team stats not found")
        
        # 월별 성적
        monthly_performance = db.execute_query("""
            SELECT 
                EXTRACT(MONTH FROM g.game_date) as month,
                COUNT(*) as games_played,
                COUNT(*) FILTER (WHERE 
                    (g.home_team_id = %s AND g.home_score > g.away_score) OR
                    (g.away_team_id = %s AND g.away_score > g.home_score)
                ) as wins,
                COUNT(*) FILTER (WHERE 
                    (g.home_team_id = %s AND g.home_score < g.away_score) OR
                    (g.away_team_id = %s AND g.away_score < g.home_score)
                ) as losses,
                COUNT(*) FILTER (WHERE g.is_draw) as draws,
                AVG(CASE WHEN g.home_team_id = %s THEN g.home_score ELSE g.away_score END) as avg_runs_scored,
                AVG(CASE WHEN g.home_team_id = %s THEN g.away_score ELSE g.home_score END) as avg_runs_allowed
            FROM games g
            WHERE (g.home_team_id = %s OR g.away_team_id = %s)
              AND EXTRACT(YEAR FROM g.game_date) = %s
              AND g.game_status = 'completed'
            GROUP BY EXTRACT(MONTH FROM g.game_date)
            ORDER BY month
        """, (team_id, team_id, team_id, team_id, team_id, team_id, team_id, team_id, season))
        
        # 최근 10경기 폼
        recent_form = db.execute_query("""
            SELECT 
                g.game_date,
                CASE 
                    WHEN (g.home_team_id = %s AND g.home_score > g.away_score) OR
                         (g.away_team_id = %s AND g.away_score > g.home_score) THEN 'W'
                    WHEN g.is_draw THEN 'D'
                    ELSE 'L'
                END as result
            FROM games g
            WHERE (g.home_team_id = %s OR g.away_team_id = %s)
              AND g.game_status = 'completed'
            ORDER BY g.game_date DESC
            LIMIT 10
        """, (team_id, team_id, team_id, team_id))
        
        return {
            "team_stats": basic_stats,
            "monthly_performance": monthly_performance,
            "recent_form": [game['result'] for game in recent_form],
            "recent_games": recent_form
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/{team_id}/rankings")
async def get_team_rankings(team_id: int, season: int = Query(2025, description="시즌 연도")):
    """팀의 각종 순위"""
    try:
        # 리그 내 각 항목별 순위
        rankings = db.execute_query("""
            WITH league_rankings AS (
                SELECT 
                    s.team_id,
                    t.team_abbreviation,
                    s.league,
                    s.wins,
                    s.losses,
                    s.win_percentage,
                    s.runs_scored,
                    s.runs_allowed,
                    s.run_differential,
                    ROW_NUMBER() OVER (PARTITION BY s.league ORDER BY s.wins DESC) as wins_rank,
                    ROW_NUMBER() OVER (PARTITION BY s.league ORDER BY s.win_percentage DESC) as win_pct_rank,
                    ROW_NUMBER() OVER (PARTITION BY s.league ORDER BY s.runs_scored DESC) as offense_rank,
                    ROW_NUMBER() OVER (PARTITION BY s.league ORDER BY s.runs_allowed ASC) as defense_rank,
                    ROW_NUMBER() OVER (PARTITION BY s.league ORDER BY s.run_differential DESC) as run_diff_rank
                FROM standings s
                JOIN teams t ON s.team_id = t.team_id
                WHERE s.season_year = %s
            )
            SELECT * FROM league_rankings WHERE team_id = %s
        """, (season, team_id))
        
        if not rankings:
            raise HTTPException(status_code=404, detail="Team rankings not found")
        
        return {
            "team_id": team_id,
            "season": season,
            "rankings": rankings[0]
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")